### DISCUSSION TITLE
Managing own framework packages development and composer

### DISCUSSION THREAD

**Discussion by exteon** [composer/composer#10949] [General]
Hi all,
I have an in-house framework for website development (much like symfony, if you will); it comprises of a number of packages (llibraries) that are reused in multiple projects.
The typical scenario is that during the development of a project, some of the libraries need to have functionality added or bugs fixed triggered by the development of the project. These modifications are usually done in-place in the project source tree.
The problem then is pushing all these project-triggered changes back to a central repository for merging and versioning.

The mechanism I used so far was to use "path" repos which would symlink dependencies to a central local path-type source repository. This allows the convenience of modifying libraries in-place and also having a central always up-to-date replica.

This mechanism works great when working on a single project at a time. I don't have that luxury anymore, and as soon as multiple projects are involved sharing the same packages with shared source code, things break loose: obviously, having the changes to a library in one project silently permeate another project which was just tested for deployment is a damning scenario.

What I envision would be of great help here is that with VCS-type repositories (I'm using git and mercurial), composer would actually create an initialized repo in `vendor/some/lib`. Then I could just commit the changes locally and push to the central repository without this affecting any other unsuspecting project. Obviously I could to this by hand for each project/dependency, but keeping track of all dependencies and initializing repositories by hand is a consuming and error-prone process.

This is so obvious and so obviously missing though that I think there may be other solutions to my problem. So I'm asking you what's the way to achieve what I need, namely being able to modify dependency packages in-place and then to easily propagate the changes to the central VCS repository for every package.

Thanks!

**Comment by Seldaek** [comment]
Set those packages to prefer installing from `source` and you'll get git clones in your vendor dir that you can work with https://getcomposer.org/doc/06-config.md#preferred-install

**Reply by exteon** [reply]
Wow, I so totally missed this! I suspected it was too obvious a feature to be missing... Thanks!

**Reply by exteon** [reply]
Ok, after trying this I bumped into another problem that I didn't have before with path repos.

I have some packages that have a commit on the (single) dev branch behind the last version tag. The best example is a library that is in development and is not yet tagged in the VCS:
```
Root composer.json requires exteon/web3-module-financial ^0.1.0, found exteon/web3-module-financial[dev-default] but it does not match your constraint
```
The package does have the proper composer.json commited with v 0.1.0. I did set minimum-stability requirement to dev.

Is there a way to pull the tips of the branches without actually altering the project's dependency specs? I.e. that composer should find that [dev-default] matches [^0.1.0] based on the composer.json and would pull and lock in dev-default@commit-id

Thanks!

**Reply by stof** [reply]
this is what branch aliases are for: https://getcomposer.org/doc/articles/aliases.md

**Reply by exteon** [reply]
Sweet! Gotta work on my documentation reading skills it seems. Thanks @stof

**Reply by exteon** [reply]
One more question pops to mind: in the deployment pipeline, `minimum-stability` will switch to `stable`. Is there a mechanism to check the Bill-Of-Materials before and after and make sure they are the same? Or is there a process that makes this unnecessary? I'm thinking that errors may happen when turning `-dev` off such as jumping to an anterior tagged commit.

**Reply by exteon** [reply]
I thought of it some more and it appears even more complicated: `minimum-stability` is a global flag that also affects other dependencies with impredictable results, so I can't use that... what I imagine as a workflow is:

1. Suffix dependencies with `@dev`. But not just root package dependencies, any co-dependency must be `@dev`-ed in the root `composer.json`. 
2. If I add `"my/x": "@dev"` in the root `composer.json` but during the development `my/x` is no longer needed in subpackages, I need to know that and remove it from root deps. Especially since `my/x` is actually loaded by a plugin much like symfony bundles and it does produce side effects by just being required.
3. When development is done, remove `@dev`, run `composer update` and somehow make sure no head has switched branch or regressed

Am I missing something here?

**Reply by Seldaek** [reply]
You could combine minimum stability dev and prefer-stable:true to avoid everything upgrading to dev

Then if something requires unreleased patches you would bump the requirement e.g lib/a latest release is 1.2.3 and it has some fixes you need in dev-main which has a branch-alias of 1.2.x-dev or 1.x-dev.. If lib/b now requires preemptively lib/a ^1.2.4@dev (or skip the @dev if you have minimum stability dev everywhere) then the root project requiring lib/b only will have to upgrade lib/a to dev-main until 1.2.4 is released at which point it'll go back to stable releases.

That may be the best compromise here..

**Reply by exteon** [reply]
Hm... Still looks so complicated for a scalable process... Add to this that VCS repos is sooo excruciatingly slow for updates...

I am now considering whether inline aliases with a gitflow approach would work better...

**Reply by Seldaek** [reply]
I mean you can still use dist by default and use `composer reinstall foo/bar --prefer-source` to switch one to a git clone if you need to work on it in-line.

**Reply by exteon** [reply]
I have to wrap my head around this, but would it work if:
```
composer reinstall my/x --prefer-source    # composer.json is not changed, source preference is volatile
composer update my/y                              # has a dependency to my/x
```
Either way, the problem is that sometimes a larger number of packages need to be changed inline... it's not just for a quick bugfix.

**Reply by Seldaek** [reply]
It would "work" in the sense that the pkg or packages installed from source remains so locally (this is only persisted to vendor dir, not in composer.json nor lock file).. But I'm not sure that fixes any of your requirement problems

**Reply by exteon** [reply]
It's kinda weird that in git repos, even where there's a single branch and I'm updating to the last commit, the HEAD is detached. This is especially annoying i.e. in PhpStorm where it won't show outgoing changes in the push dialog.

In a similar line of thought, for mercurial repositories, it would be nice if it would forward one commit if that commit is the tip and it only changes .hgtags (I know, the mercurial confusing way of dealing with tags...)

**Reply by exteon** [reply]
When updating source-mode git repositories, is `git remote show origin` really necessary? It adds about 30 seconds for me for an `update` that would otherwise be quick enough, as the repositories are all local filesystem clones

**Reply by Seldaek** [reply]
If we don't `git remote show origin` we do not detect when the default branch changes on the remote, which can lead to inconsistent state.

**Reply by exteon** [reply]
This sounds kind of cryptic to me... And isn't it a bit arbitrary? Why remote and not remote's remote? And what if origin has been abandoned 2 years ago when I made my clone and now for some reason it changes default branch silently generating side effects? I'm not sure I understand this but my instinct screams this feature should be strictly opt-in...

**Reply by Seldaek** [reply]
Look this was fixing some issues we had, but anyway I slowly have a feeling here like you are somehow affected by all sorts of problems nobody else has. So maybe everyone else is wrong, or mayyybe you need to look at your processes?

**Reply by exteon** [reply]
Yea, sorry to be a pest, I don't have a process ATM, I'm trying to figure out one. And I'd be willing to do any changes myself, if to benefit no one else but myself :) Like adding a config switch to trigger off this behavior. I'm just looking for information here trying to set up a solid process. I don't know how everyone else does it :) As for having problems noone else has, that's great actually, I remember a time when a package manager for PHP was a problem that noone had, or strict typing, or a testing automation platform :) Thanks for the help, hope I didn't trigger you too badly.

**Reply by Seldaek** [reply]
Yeah it's nice that you are willing to help, but config switches for one person do carry a maintenance cost for me, so I'm not so keen on adding those. The question is why does this take 30sec IMO.. because that doesn't sound normal.

**Reply by exteon** [reply]
Hm, good question... I was actually timing `GitDriver::getRootIdentifier`, which totals to a whooping 19.08 seconds today (faster than yesterday...). But,
```
time git remote show origin
real    0m0,007s
user    0m0,003s
sys     0m0,005s
```
Something doesn't add up here... I'll dig some more to see how it jumps to 19s.

There are 13 repos, all hosted on github, that are adding up to this.

Narrowed it down to `ProcessExecutor::execute`, all the time is spent there, altough from shell the command is lightning fast. I'll try to dig into it deeper this afternoon.

**Reply by exteon** [reply]
Ok, so the 7ms time is on the `vendor/...` clone but composer runs it on the repository clone... I ignored this. Run against the repository clone, I get, on my dev laptop and on a deployment server, similar results: `time git remote show origin`: for a SSH clone, ~1.5s, for a HTTPS clone, ~0.5s:

```
git clone https://github.com/exteon/class-helper.git && cd class-helper && time git remote show origin
real    0m0.519s
user    0m0.118s
sys     0m0.144s
```

**Reply by exteon** [reply]
I've tracked the changes and the tickets associated with the feature and I think the problem stems from this:

1. The purpose of `git remote show origin` is to keep a **cache clone** consistent with the remote repo
2. As per `GitDriver::initialize()`:
```php
        if (Filesystem::isLocalPath($this->url)) {
            $this->repoDir = $this->url;
        } else {
            $this->repoDir = $this->config->get('cache-vcs-dir') . '/' . Preg::replace('{[^a-z0-9.]}i', '-', $this->url) . '/';
```
a cache repo is NOT created for local path repos
3. In `Git::getMirrorDefaultBranch` we have:
```php
            if ($isLocalPathRepository) {
                $this->process->execute('git remote show origin', $output, $dir);
            } else {
                'git remote set-url origin -- %s && git remote show origin && git remote set-url origin -- %s'
```
I think here the first branch is oblivious of the fact that if `$isLocalPathRepository` then the repo is **not a cached clone**. So I think if we remove the ```if ($isLocalPathRepository) { $this->process->execute('git remote show origin', $output, $dir); }``` case, we keep the intended purpose of #10687 without breaking performance and consistency for local-path repos, which are NOT strictly cache clones, but can be authoritative in their own right (i.e. a fork of a remote project, a dependency archive etc).

So this would make the behavior consistent, not require a config switch, and the only question is if someone relies on the first-branch behavior of `Git::getMirrorDefaultBranch()`. Would this be fixable? cc: @glaubinix

**Reply by Seldaek** [reply]
@exteon yeah that seems like a correct analysis, fixed in https://github.com/composer/composer/pull/10996

**Reply by exteon** [reply]
Thanks, I was already testing the same change for the past week :)
I was also testing https://github.com/exteon/composer/commit/9536b43c292f0a3ffe5e826049cbeec575fe55d5 to slightly speed up things and keep backups slim (I regularly back up `~`). Seems stable.
So far I made it to ~7 seconds for an update (with all library packages fetched in source mode from a local clone which then dumps into upstream [manually for now]). The time includes syncing regular composer packages, asset packagist packages, and running plugin post-update hooks. It's a huge improvement from where I started (30 seconds just own repositories fetching). Switched all mercurial repos to git in the process (it seemed easier to do than fix `composer status` for mercurial and/or adding a non-remote-roundtrip `hg outgoing` command, which otherwise takes forever).

All in all, things are progresssing to where I think I'll have a workable process...

**Reply by Seldaek** [reply]
Ah yes that other commit looks sensible too at a glance, feel free to send a PR with that.

**Reply by exteon** [reply]
I've run into the following dilemma: for local project `my/p` I have dependency `my/x` cloned from `dev-master`. In my local project, I want to add a new dependency `some/dep` to `my/x`. I haven't found a way for composer to acknowledge the local repo's `vendor/my/x/composer.json`, unless I commit it and push it to `origin/master`. Is there a more immediate way to change and test source package deps without cluttering `origin`? Thanks...

**Reply by stof** [reply]
have you tried using a `path` repository to handle the usage of your locally modified clones ?

**Reply by exteon** [reply]
@stof I tried to use a `path` repo with `"url":"vendor/my/x"` but that didn't work as composer will delete `vendor/my/x` on the subsequent update. I'm trying to avoid creating arbitrary clones since I want to create a clear, documentable development process, part of that being a tool that tracks "dirty" dependencies in local projects so that if I come back in a month to work on `my/p` I have a tool that tells me where I left off... "Create a clone of my/x wherever" or "Track a wherever clone of my/x" kinda bumps heads with a predictable process and the tool...

**Reply by exteon** [reply]
I mean, I could make a composer plugin kind-of tool that would implement some command like `composer detach my/x` which would move `vendor/my/x` to `local-vendor/my/x` and add a path repo for `local-vendor/*`, but I was hoping there's already some easier thing that I didn't see.

**Reply by exteon** [reply]
When I'm done with this I'm publishing the docs and tools and call it `phpflow` :)
